src/trace-plugin-loader.ts

```diff
@@ -16,278 +16,485 @@
-    for (const file of Object.keys(require.cache)) {
-      if (file.match(regex)) {
-        logger_.error(
-            moduleName + ' tracing might not work as ' + file +
-            ' was loaded before the trace agent was initialized.');
-        break;
+  isSupported(version: string): boolean {
+    // The plugin is lazily loaded here.
+    const plugin = this.getPluginExportedValue();
+    // Count the number of Patch/Intercept objects with compatible version
+    // ranges
+    let numFiles = 0;
+    plugin.forEach(instrumentation => {
+      const postLoadVersions = instrumentation.versions;
+      if (!postLoadVersions || semver.satisfies(version, postLoadVersions)) {
+        numFiles++;
       }
+    });
+    // We consider a module to be unsupported if there are no Patch/Intercept
+    // objects with compatible version ranges at all.
+    return numFiles > 0;
+  }
+
+  unapplyAll() {
+    this.unpatchFns.reverse().forEach(fn => fn());
+    this.unpatchFns.length = 0;
+    this.traceApiInstances.forEach(traceApi => traceApi.disable());
+    this.traceApiInstances.length = 0;
+  }
+
+  applyPlugin<T>(moduleExports: T, file: string, version: string): T {
+    // Pre-compute a string used in logs for code clarity.
+    const logString = `${this.name}@${version}${file ? `:${file}` : ''}`;
+    // Get the exported value of the plugin value (loading it if it doesn't
+    // exist)
+    const plugin = this.getPluginExportedValue();
+    // Get a list of supported patches. This is the subset of objects in the
+    // plugin exported value with matching file/version fields.
+    const supportedPatches = plugin.filter(
+        instrumentation =>
+            semver.satisfies(version, instrumentation.versions || '*') &&
+            (file === instrumentation.file ||
+             (!file && !instrumentation.file)));
+    if (supportedPatches.length > 1) {
+      this.logger.warn(`PluginWrapper#applyPlugin: [${
+          logString}] Plugin has more than one patch/intercept object for this file. Applying all.`);
     }
+
+    // Apply each patch object.
+    return supportedPatches.reduce<T>((exportedValue, instrumentation) => {
+      // TODO(kjin): The only benefit of creating a new TraceAgent object per
+      // patched file is to give us granularity in log messages. See if we can
+      // refactor the TraceAgent class to avoid this.
+
+      this.logger.info(
+          `PluginWrapper#applyPlugin: [${logString}] Applying plugin.`);
+      if (instrumentation.patch) {
+        instrumentation.patch(
+            exportedValue, this.createTraceAgentInstance(logString));
+        // Queue a function to run if the plugin gets disabled.
+        if (instrumentation.unpatch) {
+          this.unpatchFns.push(() => {
+            this.logger.info(
+                `PluginWrapper#unapplyAll: [${logString}] Unpatching file.`);
+            instrumentation.unpatch!(exportedValue);
```

Comment:
The compiler is actually correctly detecting a nullable error here. If the `unpatch` field is cleared (or changed) after `applyPlugin()` is called, this code will do a wrong thing. You should create a local variable that is closured over from this code.
```ts
if (instrumentation.unpatch) {
  const unpatch = instrumentation.unpatch;
  this.unpatchFns.push(() => {
    // ...
    unpatch(exportedValue);
  });
}
```

Reply:
Good idea, done